File: 2019/intcode/vm.rb
class Machine
  def initialize(tape, md = nil)
    @cin, @cout, @mode = [], [], md
    @xvs, @idx, @rbx, = tape.clone, 0, 0
    @cin << @mode if @mode.is_a? Numeric
  end
  def execute
    op = @xvs[@idx].to_s.rjust(5, '0').chars
    op_code, @im = op.pop(2).join.to_i,
      op.reverse.map(&:to_i)
    @params = @xvs[@idx+1, 3]
    return if @rbx < 0
    puts "#{@xvs.take 18}"
    puts "op: #{op_code}, idx: #{@idx}, rbx: #{@rbx}, #{@params}"
      case op_code
    when 9
      @rbx += pts(1); @idx += 2
    when 8
      inst(3) {pts(1) == pts(2) ? 1:0}
    when 7
      inst(3) {pts(1) < pts(2) ? 1:0}
    when 6
      @idx = pts(1).zero? ?
        pts(2) : @idx + 3
    when 5
      @idx = pts(1).zero? ?
        @idx + 3 : pts(2)
    when 4
      # puts pts(1)
      @cout << pts(1); @idx += 2
    when 3
      inst(1) {@cin.shift}
    when 2
      inst(3) {pts(1) * pts(2)}
    when 1
      inst(3) {pts(1) + pts(2)}
    else return
  end end
  def program(*cin)
    @cin += cin; while execute
      break if @mode && !@cout.empty?
    end; @mode == false ? @cout : @cout.pop
  end
  def inst(i)
    @xvs[pts(i, ?o)] = yield
    @idx += i + 1
  end
  def pts(i, set = nil)
    v = @params[i -= 1]
    case @im[i]
    when 2; v += @rbx
      set ? v: @xvs[v]
    # when 1; set ? : v;
    when 0; set ? v : 
      @xvs[v]; else; v
  end.to_i end
  def end?
    99 == @xvs[@idx]
  end
end
# Pattern Scan & Force Floor
xvs = File.read('ins.txt').split(',').map(&:to_i)
puts Machine.new(xvs, false).program.map{|x|
  [255, x].min.chr}.join

File: 2019/Day 13/vm.py
class Program:
    def __init__(s, xvs):
        s.Idx, s.Rbx = 0, 0
        s.Tape = xvs.copy()
        s.Inputs = []
    def Push(s, Value):
        s.Inputs.append(Value)
    def NextOpCode(s):
        Code = str(s.Tape[s.Idx]).rjust(5, '0')
        s.Mode = (*map(int, Code[:-2][::-1]),)
        return getattr(s, '_' + Code[-2:])()
    def Eval(s):
        Value = True
        while Value != False:
            if Value := s.NextOpCode():
                yield Value
    def __getitem__(s, x):
        Value = s.Tape[s.Idx + x]
        if not(Type := s.Mode[x - 1]):
            return s.Tape[Value]
        if Type == 1:
            return Value
        if Type == 2:
            return s.Tape[s.Rbx + Value]
    def __setitem__(s, x, v):
        Value = s.Tape[s.Idx + x]
        if not(Type := s.Mode[x - 1]):
            s.Tape[Value] = v
        if Type == 1:
            s.Tape[s.Idx + x] = v
        if Type == 2:
            s.Tape[s.Rbx + Value] = v
    def _01(s):
        s[3] = s[1] + s[2]
        s.Idx += 4
    def _02(s):
        s[3] = s[1] * s[2]
        s.Idx += 4
    def _04(s):
        V = s[1]
        s.Idx += 2
        return V
    def _99(s):
        return False
print([*Program([21101, 55, 55, 0, 4, 0, 99]).Eval()])
